docs/topics/request-response.rst

```diff
@@ -341,6 +342,157 @@ achieve this by using ``Failure.request.cb_kwargs``::
             main_url=failure.request.cb_kwargs['main_url'],
         )
 
+
+.. _request-fingerprints:
+
+Request fingerprints
+--------------------
+
+There are some aspects of scraping, such as filtering out duplicate requests
+(see :setting:`DUPEFILTER_CLASS`) or caching responses (see
+:class:`~scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware`), where you
+need the ability to generate a short, unique identifier from a
+:class:`~scrapy.http.Request` object: a request fingerprint.
+
+You often do not need to worry about request fingerprints, the default request
+fingerprinter works for most projects.
+
+However, there is no universal way to generate a unique identifier from a
+request, because different situations require comparing requests differently.
+For example, sometimes you may need to compare URLs case-insensitively, include
+URL fragments, exclude certain URL query parameters, include some or all
+headers, etc.
+
+To change how request fingerprints are built for your requests, use the
+:setting:`REQUEST_FINGERPRINTER_CLASS` setting.
+
+.. setting:: REQUEST_FINGERPRINTER_CLASS
+
+REQUEST_FINGERPRINTER_CLASS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Default: ``'scrapy.utils.request.RequestFingerprinter'``
+
+The import path of a :ref:`request fingerprinter
+<custom-request-fingerprinter>`.
+
+The default request fingerprinter takes into account a canonical version
+(:func:`w3lib.url.canonicalize_url`) of
+:attr:`request.url <scrapy.http.Request.url>` and the values of
+:attr:`request.method <scrapy.http.Request.method>` and
+:attr:`request.body <scrapy.http.Request.body>`. It then generates an
+`SHA1 <https://en.wikipedia.org/wiki/SHA-1>`_ hash from their concatenation.
+
+
+.. _custom-request-fingerprinter:
+
+Writing your own request fingerprinter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A request fingerprinter is a class that must implement the following method:
+
+.. method:: fingerprint(self, request)
+
+   Return a :class:`str` that uniquely identifies *request*.
```

Comment:
I wonder if that's our last chance to solve https://github.com/scrapy/scrapy/pull/3420 in a backwards compatible way, or at least progress towards it. Should we require fingerprinter.fingerprint return bytes?